### ISSUE TITLE
Broken sampleRateShading

### ISSUE THREAD

**Issue by Novum** [KhronosGroup/MoltenVK#1656]
Setting `VkPipelineMultisampleStateCreateInfo::sampleShadingEnable = VK_TRUE;` and `VkPipelineMultisampleStateCreateInfo::minSampleShading = 1.0f;` should enable supersampling, but it does not. It's just regular multisampling.

Either this is a bug and it should be fixed or MoltenVK shouldn't claim to support `sampleRateShading`.

**Comment by billhollings** [comment]
The `VkPhysicalDeviceFeatures::sampleRateShading` feature enables two capabilities in Vulkan:

- >If sample shading is enabled, an implementation must invoke the fragment shader at least max(VkPipelineMultisampleStateCreateInfo::minSampleShading × VkPipelineMultisampleStateCreateInfo::rasterizationSamples, 1) times per fragment _(Vulkan spec)_.
- >`sampleRateShading` [...clip...] This also specifies whether shader modules can declare the `SampleRateShading` capability _(Vulkan spec)_.

Metal supports the second capability (super-sampling in the presence of Per-Sample builtins in the shader), but not the first (forced supersampling without such builtins):

>The implementation may decide to only execute the code that depends on the per-sample values to execute per-sample and the rest of the fragment function may execute per-fragment _(Apple MSL doc)_.

If `minSampleShading` is not zero, MoltenVK already actually attempts to declare no-op usage of Per-Sample builtins in the MSL shader, in an attempt to force that first requirement, but unfortunately Metal optimizes them away and does not run any part of the shader Per-Sample.

Disabling `VkPhysicalDeviceFeatures::sampleRateShading` would fix the app expectations for the first Vulkan requirement above, but unfortunately, per the second Vulkan requirement above, would also set an app expectation that Per-Sample builtins cannot be used in a shader.

I've opened a [discussion](https://github.com/KhronosGroup/Vulkan-Portability/issues/60) in the Portability extension about adding a `VkPhysicalDevicePortabilitySubsetFeaturesKHR::supportMinSampleShading` capability, that would allow an app to query this functionality.

Finally, to help possibly justify to Apple that the behavior should be changed, what is your use-case for wanting supersampling when the shader does not make effective use of the Per-Sample builtins?

**Comment by Novum** [comment]
Thanks for the detailed response, I wondered if it's something Metal can't support.

**Comment by billhollings** [comment]
> The use case is to enable SSAA without having to do anything special with the shaders. I maintain vkQuake for which there is more than enough GPU power to just brute force SSAA on everything.

And you find a noticeable visual difference with SSAA over MSAA in that situation? More than just the edge anti-aliasing that MSAA is focused on? Where does SSAA show better visuals over MSAA in your environment?

I'm curious, because it would help validate a possible request for Apple to support forced SSAA.

**Comment by Novum** [comment]
Yes of course there is a noticable difference. Shading aliasing is a thing, even more so with modern PBR techniques. Most people use temporal super sampling to deal with that, but that has it's own quality problems (mainly ghosting).

**Comment by billhollings** [comment]
@alyssarosenzweig suggests a fix [here](https://github.com/KhronosGroup/Vulkan-Portability/issues/60#issuecomment-1939684003).
